File: Leetcode100/leetcode100_python/MaximalSquare.py
from typing import List

class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not matrix:
            return 0
        m = len(matrix)
        n = len(matrix[0])
        dp = [[0 for _ in range(n)] for __ in range(m)]
        
        result = 0
        for i in range(m):
            for j in range(n):
                if matrix[i][j] == '1':
                    if i == 0 or j == 0:
                        dp[i][j] = 1
                    else:
                        dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1
                    result = max(result, dp[i][j])
        return result ** 2


if __name__ == "__main__":
    n = int(input())
    matrix = []
    for i in range(n):
        matrix.append([ch for ch in input().split()])
    solution = Solution()
    print(solution.maximalSquare(matrix))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/maximal_square/MaximalSquare.java
package main.java.com.jiang.leetcode100.maximal_square;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Solution {
    public int maximalSquare(char[][] matrix) {
        int m = matrix.length;
        if(m == 0) return 0;
        int n = matrix[0].length;
        int[][] dp = new int[m][n];
        int result = 0;
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(matrix[i][j] == '1') {
                    if(i == 0 || j == 0) dp[i][j] = 1;
                    else {
                        dp[i][j] = Math.min(Math.min(dp[i - 1][j], dp[i][j - 1]), 
                            dp[i - 1][j - 1]) + 1;
                    }
                    if(dp[i][j] > result) result = dp[i][j];
                }
            }
        }
        return result * result;
    }
}

public class MaximalSquare {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        List<String> inputs = new ArrayList<>();
        while(!scanner.hasNext("#")) {
            inputs.add(scanner.nextLine());
        }
        scanner.close();

        char[][] matrix = new char[inputs.size()][inputs.get(0).length()];
        for(int i = 0; i < inputs.size(); i++) {
            matrix[i] = inputs.get(i).toCharArray();
        }

        System.out.println(new Solution().maximalSquare(matrix));
    }
}